rfcs/040-tei_adapter/README.md

```diff
@@ -0,0 +1,109 @@
+# TEI Adapter
+
+## What is TEI
+
+TEI is an XML standard to represent any source in any script or language. 
+It's used at Wellcome to describe manuscripts in non latin languages.
+The TEI files are in a [GitHUb Repository](https://github.com/wellcomecollection/wellcome-collection-tei). At the 
+time of writing in April 2021 there are 622 TEI files, but the collection of manuscripts is much larger, around 10k.
+
+Currently, all information stored in TEI files is not surfaced anywhere except the GitHub repository. 
+Many of the TEI files have a dummy Sierra record that they link to, but not all
+## How we do it
+### Do we need an adapter?
+
+We need to be able to propagate updates to TEI files as they happen through to the API, and we also need to be able to
+reindex when we make changes in the pipeline. 
+
+GitHub is a relatively reliable and low latency source, so we could in theory just send the changes to a TEI transformer 
+and do a reindex by cloning the repository and/or parsing the tree of the TEI files. 
+There are some disadvantages to this approach:
+- We need a "special" reindexer for TEI.
+- We can't keep track of deleted files. If something is deleted, the change will go through the pipeline once, 
+  but there's no way of "remembering" it.
+- We need to assign versions to works resulting from TEI files in order to ensure subsequent updates are 
+  processed correctly in the pipeline. This versioning logic would have to be added explicitly to the 
+  TEI transformer whereas having an adapter and an intermediate store would allow us to use the already existing versioning
+  logic in the versioned store library.
+- Currently, we compare the number of record in the source table to the number of works in the transformer pipeline store
+  to track the success of a reindex. We wouldn't be able to do that without a store.
+
+For these reasons the proposal is to have a TEI adapter that listens to changes to TEI files and a TEI store.
+
+However, since GitHub is a relatively reliable and low latency source, we  could only store the HTTP URLs
+to the TEI files in a TEI store, and not the full XML. This has the advantage that we don't have to manage an 
+ever-growing S3 bucket (we don't have a way of cleaning up old versions of a record), but it has some disadvantages:
+- GitHub outages (which are rare but happen) might affect our reindexes.
+- We might be throttled during a reindex
+- We go to the public internet for every TEI file for every reindex as opposed to getting each file exactly once 
+  per every change that affects it.
+  
+For these reasons, the proposal here is to store the whole TEI file in a full VHS.
+### Store structure
+
+Therefore, taking as an example
+[WMS_Arabic_529](https://github.com/wellcomecollection/wellcome-collection-tei/blob/master/Arabic/WMS_Arabic_529.xml), 
+the proposed store record would be:
+
+| Id | Payload | deleted  | version |
+| --- | ------- | -------| --------|
+| WMS_Arabic_529 | ... | false | 1 |
```

Comment:
Presumably the payload would be an object in S3?

You'll need to change the stores somewhat (or create your own) to put that deleted metadata in DynamoDB.

Reply:
hmm good point, let me see how it's done in calm

Reply:
So the calm vhs has a isDeleted flag in the dynamo table, so I think we can steal whatever code the Calm adapter uses for that.

Yes the payload is the location in s3. I'll write it better